sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_ptr.asciidoc

```diff
@@ -0,0 +1,510 @@
+
+:blank: pass:[ +]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+// This is necessary for asciidoc, but not for asciidoctor
+:cpp: C++
+:dpcpp: DPC++
+
+== Introduction
+This extension introduces a pointer wrapper class that provides a mechanism to
+attach compile-time constant information to a pointer in a manner that allows
+the compiler to reliably maintain and analyze the information.
+
+== Notice
+
+Copyright (c) 2021 Intel Corporation.  All rights reserved.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are
+trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc.
+used by permission by Khronos.
+
+== Status
+
+Working Draft
+
+This is a preview extension specification, intended to provide early access to
+a feature for review and community feedback. When the feature matures, this
+specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are
+subject to change they are not intended to be used by shipping software
+products.
+
+== Version
+
+Built On: {docdate} +
+Revision: 2
+
+== Contact
+
+Abhishek Tiwari, Intel (abhishek2 'dot' tiwari 'at' intel 'dot' com)
+
+== Contributors
+
+Abhishek Tiwari, Intel +
+Gregory Lueck, Intel +
+Jason Sewall, Intel +
+Jessica Davies, Intel +
+Joe Garvey, Intel +
+John Pennycook, Intel +
+Michael Kinsner, Intel +
+Roland Schulz, Intel
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, revision 4.
+
+It also depends on the `sycl_ext_oneapi_properties` extension.
+
+== Overview
+
+[NOTE]
+====
+In this document, we use the shortened form `annotated_ptr` to refer to the
+proposed `sycl::ext::oneapi::annotated_ptr` class.
+====
+
+The purpose of this document is to clearly describe and specify `annotated_ptr`
+and related concepts, types, and mechanisms, and to give examples and context
+for their usage.
+
+=== Motivation
+
+When compiling a program with pointer operations, it is often desirable to
+attach certain information such as optimization hints or functional directives
+onto the pointer itself.
+
+Any acceptable mechanism to attach such information should meet the following requirements:
+  1. It should not be burdensome for users to use the mechanism.
+  2. It should be reasonably scalable and not add significant burden on compiler developers to maintain the mechanism.
+  3. It should preserve the attached information through most uses of the pointer.
+
+For example, consider the following case: We want to specify some compile-time constant information `I` for a pointer. One possible method to do so is for the compiler to provide a special function for reading to and writing from the pointer and have this function convey `I` to the compiler at the call site where the read or write occurs.
```

Comment:
To aid with future review and changes, we typically recommend wrapping long lines at 80 columns.